### ISSUE TITLE
Function Lessons Reconfiguration

### ISSUE THREAD

**Issue by flannery-denny** [bootstrapworld/curriculum#797]
- [ ] Functions as Machines - this builds directly on the defining values lesson, introducing a very narrow function concept to start. It's most of our regular gt, bc, etc as labor-saving devices lesson, emphasizing the progression from gt50 = triangle(50, "solid", "green") to gt(50) is triangle(50, "solid", "green"). It's fairly rote and programming-focused to start, but there's tons of engaging problems and it drives home two critical foundations: functions are useful, and the concept of function machines. This makes it easy to cover Pyret errors too, which are also fairly rote and should be gotten out of the way quickly.
- no math
- function machine reference
- define other functions
- why are functions efficient?

- [ ] Functions as Examples and Definitions - this broadens the function concept, leveraging the existing "3 representations" lesson to get students thinking about functions in flexible ways. This lesson also introduces/formalizes function notation, since a definition is just another representation of functions.
- this lesson focuses on the math
- laying the foundation for the design recipe is not critical

- [ ] Properties of Functions - this is where I'd slot your proposed lesson, with some modifications. I think the vertical line test is great here, and this is the right place to hit domain&range hard like we've always wanted. Continuous v. Discrete is also great here, and goes naturally with D&R. Function notation is already covered elsewhere, as are Pyret errors. This lesson could lean harder on graphing functions - a known gap for us, and a natural companion to discrete v. continuous properties.

- [ ] Functions can be Linear - This one is basically our existing linearity✅, layering on straight-line relationships as another property a function can have. It also spirals back on multiple representations, and gives us a nice opportunity to test their understanding about things like the vertical line test, continuity, etc.

- [ ] Properties of Linear Functions - This is basically our existing linearity2✅, which could use a lot of improvement from the ideas you're proposing. The main thing here is that we can't introduce slope before we introduce the idea of straight-line relationships, and slope is better understood when you have the anchors of continuity and graphing.

**Comment by flannery-denny** [comment]
@schanzer In the last commit, I've split the defining functions lesson in two. 

**Functions as Machines**
- [ ] This definition stands out to me as in need of broadening _function a mathematical object that consumes inputs and produces an output _. Would you agree?

- [ ] Is there a way to allow the examples at the top of Matching Examples and Definitions to run the full width instead of breaking lines like this?
![Screen Shot 2022-03-06 at 2 32 36 PM](https://user-images.githubusercontent.com/66691020/156941084-12913ad1-60d1-4bb7-bc86-ee73f941cd18.png)

- [ ] Do you really want to have them generate functions without scaffolding as with the current list of ideas on Defining Functions or would you be ok with asking them to write examples for each of these as currently happens on [this page](https://bootstrapworld.org/materials/latest/en-us/lessons/defining-functions/pages/contracts-examples-definitions.html), perhaps formatted like this page without the contract section? 

- [ ] The suggestion was to have students generate lots of functions in the _Functions as Machines_ lesson, which I think means we should have a starter file. Would you be ok with me engaging students with this starter file that includes contracts, examples and definition, even though the core of that work is being saved for the next lesson? [gt starter file](https://code.pyret.org/editor#share=1wvO-su91pFXQJcsIAa5IUOVcUea06YuY)

**Functions as Examples and Definitions**
- [ ] The lesson you've described as _Functions as Examples and Definitions_ only refers to two representations in the title. Given its focus on math, that might make sense. But take a look at the commit to see all of the materials from the old Defining Functions lesson that connect to contracts. Wondering whether you already have a vision for which lesson these fit into? Do we need 3 lessons here? Or, related to my previous question do functions/examples/definitions/contracts actually all belong in functions as machines and what you really want here is a lesson on **Functions in Math**? Turn out I'm still not clear on where you were hoping for the line to be drawn.

**Comment by flannery-denny** [comment]
@schanzer The more I think about discrete or continuous, I think it makes the most sense to address it in the Data Science pathway.

**Comment by schanzer** [comment]
@flannery-denny I want to hear more about your thinking re:continuous v. discrete. It seems like it fits very naturally in the math pathway, and with Data Science there's a real risk of confusing scatterplots with non-continuous functions. Almost every single scatterplot kids see is _not a function_ anyway!

I like the way you were describing using the "shadowing" error message , but I'm still curious about how you intend to use the others.

I do not understand the cruise control analogy.

**Comment by flannery-denny** [comment]
@schanzer the error messages were a note to me. The comment I was asking you to pay attention to was this one from above [https://github.com/bootstrapworld/curriculum/issues/797#issuecomment-1060040234](https://github.com/bootstrapworld/curriculum/issues/797#issuecomment-1060040234)

**Comment by schanzer** [comment]
@flannery-denny ah! thanks for the correction.

1. Yes, I totally agree.
2. Not without _extensive_ CSS and build script work. The flexbox stuff will likely grow to take care of this, but that's a long way off. For now, I'd suggest hardcoding the formatting you want in each language, and wrapping them in `@ifproglang`

**Comment by flannery-denny** [comment]
![simple function machine w coe](https://user-images.githubusercontent.com/66691020/158382461-769c134e-7240-48bb-a470-7ea6ab0dc448.png)

**Comment by flannery-denny** [comment]
![simplified function machine](https://user-images.githubusercontent.com/66691020/158390845-e556d83f-adcf-464d-b38e-be2d5e872758.png)

**Comment by flannery-denny** [comment]
@schanzer Looking at the slide deck and seeing this question 

- [ ] Is there a good reason that you didn't refer explicitly to the slope-intercept form of the line when writing the linearity lessons? 
- [ ] Also, turned up this OK Standard. _Express linear equations in slope-intercept, point-slope, and standard forms and convert between these forms. Given sufficient information (slope and y-intercept, slope and one-point on the line, two points on the line, x- and y-intercept, or a set of data points), write the equation of a line._ How much do care about covering it?

**Comment by flannery-denny** [comment]
@schanzer I found a number of additional materials that needed to be moved to contracts-examples-definitions. Please confirm that the way I have integrated the additional pages works for you.

**Comment by flannery-denny** [comment]
Found this page in the Restating the Problem lesson that's gone, but also wasn't using it.
[MappingExamplesWithCoE.pdf](https://github.com/bootstrapworld/curriculum/files/8414334/MappingExamplesWithCoE.pdf)
It has the most to do with the contracts-examples-definitions lesson. 

If we're going to bring it back, I think we need to make a solution version of it that models how we would use it for one of the problems they're already writing in the lesson. 

What's your feeling on it @schanzer?
Think it's worth putting on the to do?

**Comment by schanzer** [comment]
@flannery-denny I think it's worth doing -- that worksheet led to a number of "ahas" for teachers. I've moved it to that lesson and included draft language/styling. Let me know what you think?

**Comment by schanzer** [comment]
@flannery-denny is there more to do here, or anything you're waiting on me for?
